### PR TITLE
[DSK] Implement Niko, Light of Hope

### PR DESCRIPTION
**Pull request by Grath**
Add UntilTheNextEndstep duration for 'until the beginning of the next end step' on the copy effect.

### CODE REVIEW

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java [hunk 1]
```diff
@@ -298,6 +307,7 @@ public boolean isInactive(Ability source, Game game) {
             case UntilYourNextTurn:
             case UntilEndOfYourNextTurn:
             case UntilYourNextEndStep:
+            case UntilTheNextEndStep:
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java:310]: Must use cleanup step, that place for “your” durations only:

IMG_0867

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectsList.java [hunk 2]
```diff
@@ -156,6 +156,7 @@ private boolean isInactive(T effect, Game game) {
                     case UntilEndOfYourNextTurn:
                     case UntilEndCombatOfYourNextTurn:
                     case UntilYourNextEndStep:
+                    case UntilTheNextEndStep:
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/effects/ContinuousEffectsList.java:159]: Same:

IMG_0868

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java [hunk 3]
```diff
@@ -93,7 +95,7 @@ protected ContinuousEffectImpl(final ContinuousEffectImpl effect) {
         this.startingControllerId = effect.startingControllerId;
         this.startingTurnWasActive = effect.startingTurnWasActive;
         this.effectStartingOnTurn = effect.effectStartingOnTurn;
-        this.effectStartingEndStep = effect.effectStartingEndStep;
+        this.effectControllerStartingEndStep = effect.effectControllerStartingEndStep;
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java:98]: Miss copy constructor fields — added 3 new, but copy changed for 1:

IMG_0869

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java [hunk 4]
```diff
@@ -56,9 +56,11 @@ A Characteristic Defining Ability (CDA) is an ability that defines a characteris
 
     // until your next turn or until end of your next turn
     private UUID startingControllerId; // player to check for turn duration (can't different with real controller ability)
+    private UUID activePlayerId; // Player whose turn the effect started on
     private boolean startingTurnWasActive; // effect started during related players turn and related players turn was already active
     private int effectStartingOnTurn = 0; // turn the effect started
-    private int effectStartingEndStep = 0;
+    private int effectControllerStartingEndStep = 0;
+    private int effectActivePlayerStartingEndStep = 0;
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java:63]: Use effecStartingXXX and startingActivePlayer naming

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java (hunk 1 above)
Comment (Grath) [reply Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java:310]: Cleanup step is too late, it ends 'beginning of next end step' and not cleanup step.

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectsList.java (hunk 2 above)
Comment (Grath) [reply Mage/src/main/java/mage/abilities/effects/ContinuousEffectsList.java:159]: Same

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java (hunk 4 above)
Comment (Grath) [reply Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java:63]: Done.

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java (hunk 1 above)
Comment (Grath) [reply Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java:310]: Updated comment to reflect that this section is now being used for things other than 'your turn'.

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectsList.java (hunk 2 above)
Comment (Grath) [reply Mage/src/main/java/mage/abilities/effects/ContinuousEffectsList.java:159]: This section already contains not-'your' durations (custom, for example) so I'm leaving it.

File: Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java (hunk 1 above)
Comment (JayDi85) [reply Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java:310]: That's wrong usage. Don't use conditional durations here. It's a static duration and must be clean due rules -- at the begin of end step like other static duration does. See example with end of turn and clean up:
shot_240930_222543

Implementation like that:
shot_240930_222403

Comment (Grath) [reply Mage/src/main/java/mage/abilities/effects/ContinuousEffectImpl.java:310]: Done.

**Pull request merged**